File: scripts/computer-science/pow/code.py
def get_steps(y):
   i = 0
   while True:
      if y < ( 2 << i ):
         return i - 1

      i += 1

def bin_multiply(x, y):
   # print(f'bin_multiply call - received {x} and {y}')

   if y == 0:
      return 0

   if y == 1:
      return x

   bit_shift_steps = get_steps(y)
   print(f'SHIFT STEPS: {bit_shift_steps}')

   remainder = y - ( 2 << bit_shift_steps )

   print(f'Will shift {x} by {bit_shift_steps} and add {x} times {remainder}')

   return ( x << bit_shift_steps + 1 ) + bin_multiply(x, remainder)


def multiply(x, y):
   #  print(f'Multiply call - received {x} and {y}')
    def extract_lsb(x):
        return x >> 1, x & 1

    shifted_y, lsb = extract_lsb(y)

    return x if shifted_y == 0 else ( bin_multiply(x, lsb) + multiply(x << 1, shifted_y) )

def fPow(base, exp):
   # print(f'fPow - received: {base} and {exp}')
   return base if exp == 1 else multiply(base, fPow(base, exp - 1))

base = 4
for x in [1, 2, 3, 4, 5, 6, 7, 8]:
   print(f'{base} to the power of {x} is: {fPow(base, x)}')

File: scripts/computer-science/pow/c/code.c
#include <stdio.h>

struct lsb_extract_type
{
    int shifted_y;
    int lsb;
};

int get_steps(int y){
    int i = 0;

    while (1 == 1){
        if (y < ( 2 << i )){
            return i - 1;
        }

        i += 1;
    }
}

int bin_multiply(int x, int y){
   if (y == 0){
       return 0;
   };

   if (y == 1){
       return x;       
   };      

   int bit_shift_steps = get_steps(y);
   int remainder = y - ( 2 << bit_shift_steps );
   return ( x << (bit_shift_steps + 1) ) + bin_multiply(x, remainder);
}

struct lsb_extract_type extract_lsb(int x)
{
    struct lsb_extract_type temp;
    temp.shifted_y = x >> 1;
    temp.lsb = x & 1;
    return temp;
}
       

int multiply(int x, int y){    
    
    struct lsb_extract_type lsb_extract;
    lsb_extract = extract_lsb(y);

    int result = (lsb_extract.shifted_y == 0) ? x : ( bin_multiply(x, lsb_extract.lsb) + multiply(x << 1, lsb_extract.shifted_y) );
    return result;
}

long int fPow(int base, int exp){
    long result = (exp == 1) ? base : multiply(base, fPow(base, exp - 1));
    return result;
}

int main(){
    int base = 4;
    for (int x = 1; x < 9; x++){
        printf("%d to the power of %d is: %ld\n", base, x, fPow(base, x));
    }
}